book-content/chapters/10-deriving-types.md

```diff
@@ -0,0 +1,1460 @@
+
+Another example is a union type.
+
+```typescript
+type Triangle = {
+  type: "triangle";
+  sideLength: number;
+};
+
+type Rectangle = {
+  type: "rectangle";
+  width: number;
+  height: number;
+};
+
+type Shape = Triangle | Rectangle;
+```
+
+A derived type represents a relationship. That relationship is one-way. `Shape` can't go back and modify `Triangle` or `Rectangle`. But any changes to `Triangle` and `Rectangle` will ripple through to `Shape`.
+
+When well-designed, derived types can create huge gains in productivity. We can make changes in one place and have them propagate throughout our application. This is a powerful way to keep our code DRY and to leverage TypeScript's type system to its fullest.
+
+This has tradeoffs. We can think of deriving as a kind of coupling. If we change a type that other types depend on, we need to be aware of the impact of that change. We'll discuss deriving vs decoupling in more detail at the end of the chapter.
+
+But for now, let's look at some of the tools TypeScript provides for deriving types.
+
+## The `keyof` Operator
+
+The `keyof` operator allows you to extract the keys from an object type into a union type.
+
+Starting with our familiar `Album` type:
+
+```typescript
+interface Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+}
+```
+
+We can use `keyof Album` and end up with a union of the `"title"`, `"artist"`, and `"releaseYear"` keys:
+
+```typescript
+type AlbumKeys = keyof Album; // "title" | "artist" | "releaseYear"
+```
+
+Since `keyof` tracks the keys from a source, any changes made to the type will automatically be reflected in the `AlbumKeys` type.
+
+```typescript
+interface Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+  genre: string; // added 'genre'
+}
+
+type AlbumKeys = keyof Album; // "title" | "artist" | "releaseYear" | "genre"
+```
+
+The `AlbumKeys` type can then be used to help ensure a key being used to access a value in an `Album` is valid as seen in this function:
+
+```typescript
+function getAlbumDetails(album: Album, key: AlbumKeys) {
+  return album[key];
+}
+```
+
+If the key passed to `getAlbumDetails` is not a valid key of `Album`, TypeScript will show an error:
+
+```ts twoslash
+// @errors: 2345
+function getAlbumDetails(album: Album, key: AlbumKeys) {
+  return album[key];
+}
+
+interface Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+  genre: string; // added 'genre'
+}
+
+type AlbumKeys = keyof Album; // "title" | "artist" | "releaseYear" | "genre"
+
+const album: Album = {
+  title: "Kind of Blue",
+  artist: "Miles Davis",
+  releaseYear: 1959,
+  genre: "Jazz",
+};
+
+// ---cut---
+getAlbumDetails(album, "producer");
+```
+
+`keyof` is an important building block when creating new types from existing types. We'll see later how we can use it with `as const` to build our own type-safe enums.
```

Comment:
it might be good to call out those 2 gotchas:
```ts
const oops: keyof { concrete: "foo"; [k: string]: unknown } = "whatever";
const test: keyof { [k: string]: unknown } = 42;
```